rust-runtime/smithy-json/src/serialize.rs

```diff
@@ -0,0 +1,451 @@
+    pub fn start_object(&mut self, key: &str) -> JsonObjectWriter {
+        self.key(key);
+        JsonObjectWriter::new(&mut self.json)
+    }
+
+    /// Finishes the object.
+    pub fn finish(self) {
+        self.json.push('}');
+    }
+
+    fn key(&mut self, key: &str) {
+        if self.started {
+            self.json.push(',');
+        }
+        self.started = true;
+
+        self.json.push('"');
+        self.json.push_str(&escape_string(key));
+        self.json.push_str("\":");
+    }
+}
+
+pub struct JsonArrayWriter<'a> {
+    json: &'a mut String,
+    started: bool,
+}
+
+impl<'a> JsonArrayWriter<'a> {
+    pub fn new(output: &'a mut String) -> Self {
+        output.push('[');
+        Self {
+            json: output,
+            started: false,
+        }
+    }
+
+    /// Writes a null value to the array.
+    pub fn null(&mut self) -> &mut Self {
+        self.comma_delimit();
+        self.json.push_str("null");
+        self
+    }
+
+    /// Writes the boolean `value` to the array.
+    pub fn boolean(&mut self, value: bool) -> &mut Self {
+        self.comma_delimit();
+        self.json.push_str(match value {
+            true => "true",
+            _ => "false",
+        });
+        self
+    }
+
+    /// Writes a string to the array.
+    pub fn string(&mut self, value: &str) -> &mut Self {
+        self.comma_delimit();
+        append_string(&mut self.json, value);
+        self
+    }
+
+    /// Writes a string `value` to the array without escaping it.
+    pub fn trusted_string(&mut self, value: &str) -> &mut Self {
+        self.comma_delimit();
+        append_string(&mut self.json, value);
+        self
+    }
+
+    /// Writes a number `value` to the array.
+    pub fn number(&mut self, value: Number) -> &mut Self {
+        self.comma_delimit();
+        append_number(&mut self.json, value);
+        self
+    }
+
+    /// Writes an Instant `value` with the given `key` and `format`.
+    pub fn instant(&mut self, instant: &Instant, format: Format) -> &mut Self {
+        self.comma_delimit();
+        append_instant(&mut self.json, instant, format);
+        self
+    }
+
+    /// Starts a nested array inside of the array.
+    pub fn start_array(&mut self) -> JsonArrayWriter {
+        self.comma_delimit();
+        JsonArrayWriter::new(&mut self.json)
+    }
+
+    /// Starts a nested object inside of the array.
+    pub fn start_object(&mut self) -> JsonObjectWriter {
+        self.comma_delimit();
+        JsonObjectWriter::new(&mut self.json)
+    }
+
+    /// Finishes the array.
+    pub fn finish(self) {
+        self.json.push(']');
+    }
+
+    fn comma_delimit(&mut self) {
+        if self.started {
+            self.json.push(',');
+        }
+        self.started = true;
+    }
+}
+
+fn append_string(json: &mut String, value: &str) {
+    append_trusted_string(json, &escape_string(value));
+}
+
+fn append_trusted_string(json: &mut String, value: &str) {
+    json.push('"');
```

Comment:
I really like this. Will add.

Reply:
Slight problem though: it won't actually be `Cow::Borrowed` in this case:
```
object.string("line one\nline two");
```
Since that string will feed into `escape_str` to become `"line one\\nline two"` before making its way into `append_string_unchecked`, which then asserts escaping it is a `Cow::Borrowed`, except that that won't be the case, because running escaping on it again would change it to `"line one\\\\nline two"` to preserve the two slashes.

If we really want to be safe about it, I could create an enum that tracks the "escapedness" of the string, if you think that's worth it.

Reply:
ah I see! I don't think that's worth it

Reply:
Or actually we can just not have `append_string` go through trusted string

Reply:
Yeah, that seems like a good compromise. Have some duplicated code to insert quotes for the added safety.